skfuzzy/control/controlsystem.py

```diff
@@ -428,19 +428,40 @@ def find_memberships(self):
         if len(self.var.terms.keys()) == 0:
             raise ValueError("Set term membership function(s) first")
 
+        '''
+        First we have to upsample the universe of self.var in order to add the key points of the membership function
+        based on the activation level for this consequent, using the interp_value function, which interpolates the `xx`
+        values in the universe such that its membership function value is the activation level.
+        '''
+        add_universe = set()
+        for label, term in self.var.terms.items():
+            cut = term.membership_value[self.sim]
+            if cut is None:
+                continue  # No membership defined for this adjective
+            add_xx = interp_value(self.var.universe, term.mf, cut)
+            add_universe.update(add_xx)
+        # We are only interested in points not in self.var.universe
+        add_universe = add_universe-set(self.var.universe)
+        upsampled_universe = zip(self.var.universe.tolist() + list(add_universe), range(len(self.var.universe))+[None]*len(add_universe))
+        upsampled_universe.sort(key=lambda element: element[0])
+        upsampled_mf_indices = map(lambda element: element[1], upsampled_universe)
+        upsampled_universe = np.array(map(lambda element: element[0],upsampled_universe))
+
         # Initilize membership
-        output_mf = np.zeros_like(self.var.universe, dtype=np.float64)
+        output_mf = np.zeros_like(upsampled_universe, dtype=np.float64)
 
         # Build output membership function
         term_mfs = {}
         for label, term in self.var.terms.items():
             cut = term.membership_value[self.sim]
             if cut is None:
                 continue  # No membership defined for this adjective
-            term_mfs[label] = np.minimum(cut, term.mf)
+            upsampled_mf = [term.mf[upsampled_mf_indices[i]] if upsampled_mf_indices[i] is not None else \
+                                interp_membership(self.var.universe, term.mf, upsampled_universe[i]) for i in xrange(len(upsampled_mf_indices))]
```

Comment:
I think this use of `xrange` will also have to be changed for Python 3 compatibility. The good news is that in Python 3, `range` behaves exactly like `xrange` in Python 2. So, with the import `from builtins import range`, a simple `range` should work here.